File: tablet_qt/menu/mainmenu.cpp
/*
    Copyright (C) 2012, University of Cambridge, Department of Psychiatry.
    Created by Rudolf Cardinal (rnc1001@cam.ac.uk).

    This file is part of CamCOPS.

    CamCOPS is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    CamCOPS is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CamCOPS. If not, see <https://www.gnu.org/licenses/>.
*/

#include "mainmenu.h"

#include <QDebug>
#include <QSharedPointer>

#include "common/uiconst.h"
#include "core/networkmanager.h"
#include "lib/uifunc.h"
#include "menu/addictionmenu.h"
#include "menu/affectivemenu.h"
#include "menu/alltasksmenu.h"
#include "menu/anonymousmenu.h"
#include "menu/catatoniaepsemenu.h"
#include "menu/clinicalmenu.h"
#include "menu/clinicalsetsmenu.h"
#include "menu/cognitivemenu.h"
#include "menu/eatingdisordersmenu.h"
#include "menu/executivemenu.h"
#include "menu/globalmenu.h"
#include "menu/helpmenu.h"
#include "menu/neurodiversitymenu.h"
#include "menu/patientsummarymenu.h"
#include "menu/personalitymenu.h"
#include "menu/physicalillnessmenu.h"
#include "menu/psychosismenu.h"
#include "menu/researchmenu.h"
#include "menu/researchsetsmenu.h"
#include "menu/serviceevaluationmenu.h"
#include "menu/settingsmenu.h"
#include "menulib/menuitem.h"
#include "menulib/menuproxy.h"

MainMenu::MainMenu(CamcopsApp& app) :
    MenuWindow(app, uifunc::iconFilename(uiconst::ICON_CAMCOPS), true)
{
}

QString MainMenu::title() const
{
    return tr("CamCOPS: Cambridge Cognitive and Psychiatric Assessment Kit");
}

void MainMenu::makeItems()
{
    m_items = {
        MAKE_CHANGE_PATIENT(m_app),
        MAKE_MENU_MENU_ITEM(PatientSummaryMenu, m_app),
        MenuItem(
            tr("Upload data to server"),
            std::bind(&MainMenu::upload, this),
            uifunc::iconFilename(uiconst::ICON_UPLOAD)
        )
            .setNotIfLocked(),
        MAKE_MENU_MENU_ITEM(HelpMenu, m_app),
        MAKE_MENU_MENU_ITEM(SettingsMenu, m_app),

        MenuItem(tr("Tasks by type")).setLabelOnly(),
        // Alphabetical by title in English
        MAKE_MENU_MENU_ITEM(AffectiveMenu, m_app),
        MAKE_MENU_MENU_ITEM(AnonymousMenu, m_app),
        MAKE_MENU_MENU_ITEM(CatatoniaEpseMenu, m_app),
        MAKE_MENU_MENU_ITEM(ClinicalMenu, m_app),
        MAKE_MENU_MENU_ITEM(CognitiveMenu, m_app),
        MAKE_MENU_MENU_ITEM(AddictionMenu, m_app),
        MAKE_MENU_MENU_ITEM(EatingDisordersMenu, m_app),
        MAKE_MENU_MENU_ITEM(ExecutiveMenu, m_app),
        MAKE_MENU_MENU_ITEM(GlobalMenu, m_app),
        MAKE_MENU_MENU_ITEM(NeurodiversityMenu, m_app),
        MAKE_MENU_MENU_ITEM(PersonalityMenu, m_app),
        MAKE_MENU_MENU_ITEM(PhysicalIllnessMenu, m_app),
        MAKE_MENU_MENU_ITEM(PsychosisMenu, m_app),
        MAKE_MENU_MENU_ITEM(ServiceEvaluationMenu, m_app),
        MAKE_MENU_MENU_ITEM(ResearchMenu, m_app),

        MenuItem(tr("Task collections")).setLabelOnly(),
        MAKE_MENU_MENU_ITEM(ClinicalSetsMenu, m_app),
        MAKE_MENU_MENU_ITEM(ResearchSetsMenu, m_app),
        MAKE_MENU_MENU_ITEM(AllTasksMenu, m_app),
    };
    connect(
        &m_app, &CamcopsApp::fontSizeChanged, this, &MainMenu::reloadStyleSheet
    );
}

void MainMenu::upload()
{
    m_app.upload();
}

File: defunct_tablet_titanium/Resources/menu/MainMenu.js
// MainMenu.js

/*
    Copyright (C) 2012-2016 Rudolf Cardinal (rudolf@pobox.com).
    Department of Psychiatry, University of Cambridge.
    Funded by the Wellcome Trust.

    This file is part of CamCOPS.

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
*/

/*jslint node: true, newcap: true */
"use strict";
/*global L */

module.exports = function MainMenu() {

    var MenuWindow = require('menulib/MenuWindow'),
        platform = require('lib/platform'),
        UICONSTANTS = require('common/UICONSTANTS'),
        data,
        self;

    function upload() {
        var uifunc = require('lib/uifunc');
        uifunc.upload(self.tiview);
    }

    data = [
        UICONSTANTS.CHANGE_PATIENT_MENU_LINE,
        {
            maintitle: L('menutitle_patient_summary'),
            icon: UICONSTANTS.ICON_PATIENT_SUMMARY,
            window: 'screen/PatientSummaryWindow'
        },
        {
            maintitle: L('menutitle_upload'),
            icon: UICONSTANTS.ICON_UPLOAD,
            func: upload,
            unsupported: (!platform.isDatabaseSupported)
        },
        {
            maintitle: L('menutitle_help'),
            icon: UICONSTANTS.ICON_MENU_HELP,
            window: 'menu/HelpMenu'
        },
        {
            maintitle: L('menutitle_settings'),
            arrowOnRight: true,
            icon: UICONSTANTS.ICON_MENU_SETTINGS,
            window: 'menu/SettingsMenu',
            notIfLocked: true
        },
        {
            maintitle: L('menutitle_clinical'),
            arrowOnRight: true,
            icon: UICONSTANTS.ICON_MENU_CLINICAL,
            window: 'menu/ClinicalMenu'
        },
        {
            maintitle: L('menutitle_global'),
            arrowOnRight: true,
            icon: UICONSTANTS.ICON_MENU_GLOBAL,
            window: 'menu/GlobalMenu'
        },
        {
            maintitle: L('menutitle_cognitive'),
            arrowOnRight: true,
            icon: UICONSTANTS.ICON_MENU_COGNITIVE,
            window: 'menu/CognitiveMenu'
        },
        {
            maintitle: L('menutitle_affective'),
            arrowOnRight: true,
            icon: UICONSTANTS.ICON_MENU_AFFECTIVE,
            window: 'menu/AffectiveMenu'
        },
        {
            maintitle: L('menutitle_addiction'),
            arrowOnRight: true,
            icon: UICONSTANTS.ICON_MENU_ADDICTION,
            window: 'menu/AddictionMenu'
        },
        {
            maintitle: L('menutitle_psychosis'),
            arrowOnRight: true,
            icon: UICONSTANTS.ICON_MENU_PSYCHOSIS,
            window: 'menu/PsychosisMenu'
        },
        {
            maintitle: L('menutitle_catatonia_epse'),
            arrowOnRight: true,
            icon: UICONSTANTS.ICON_MENU_CATATONIA,
            window: 'menu/CatatoniaEpseMenu'
        },
        {
            maintitle: L('menutitle_personality'),
            arrowOnRight: true,
            icon: UICONSTANTS.ICON_MENU_PERSONALITY,
            window: 'menu/PersonalityMenu'
        },
        {
            maintitle: L('menutitle_executive'),
            arrowOnRight: true,
            icon: UICONSTANTS.ICON_MENU_EXECUTIVE,
            window: 'menu/ExecutiveMenu'
        },
        {
            maintitle: L('menutitle_research'),
            arrowOnRight: true,
            icon: UICONSTANTS.ICON_MENU_RESEARCH,
            window: 'menu/ResearchMenu'
        },
        {
            maintitle: L('menutitle_anonymous'),
            arrowOnRight: true,
            icon: UICONSTANTS.ICON_MENU_ANONYMOUS,
            window: 'menu/AnonymousMenu'
        },
        {
            maintitle: L('menutitle_sets_clinical'),
            arrowOnRight: true,
            icon: UICONSTANTS.ICON_MENU_SETS_CLINICAL,
            window: 'menu/ClinicalSetsMenu'
        },
        {
            maintitle: L('menutitle_sets_research'),
            arrowOnRight: true,
            icon: UICONSTANTS.ICON_MENU_SETS_RESEARCH,
            window: 'menu/ResearchSetsMenu'
        },
        {
            maintitle: L('menutitle_alltasks'),
            arrowOnRight: true,
            icon: UICONSTANTS.ICON_MENU_ALLTASKS,
            window: 'menu/AllTasksMenu'
        }
    ];

    self = new MenuWindow({
        title: L('app_long_title'),
        backbutton: false,
        icon: UICONSTANTS.ICON_CAMCOPS,
        data: data,
        addfooter: false,
        exitOnClose: true // special for the top menu
    });

    return self;
};
